balancer-js/src/modules/swaps/joinExit/actions/swap.ts

```diff
@@ -0,0 +1,179 @@
+import { BigNumber } from '@ethersproject/bignumber';
+import { MaxUint256 } from '@ethersproject/constants';
+import { SubgraphPoolBase, SwapV2 } from '@balancer-labs/sor';
+import {
+  Relayer,
+  EncodeBatchSwapInput,
+} from '@/modules/relayer/relayer.module';
+import { FundManagement, SwapType } from '../../types';
+import { ActionStep, ActionType, SwapAction, BatchSwapAction } from './types';
+import {
+  getActionStep,
+  getActionAmount,
+  getActionMinOut,
+  getActionOutputRef,
+} from './helpers';
+
+function isBpt(pools: SubgraphPoolBase[], token: string): boolean {
+  return pools.some((p) => p.address.toLowerCase() === token.toLowerCase());
+}
+
+/**
+ * Create a SwapAction with relevant info.
+ * @param swapType
+ * @param swap
+ * @param mainTokenInIndex
+ * @param mainTokenOutIndex
+ * @param opRefKey
+ * @param assets
+ * @param slippage
+ * @param pools
+ * @param user
+ * @param relayer
+ * @returns
+ */
+export function createSwapAction(
+  swap: SwapV2,
+  mainTokenInIndex: number,
+  mainTokenOutIndex: number,
+  opRefKey: number,
+  assets: string[],
+  slippage: string,
+  pools: SubgraphPoolBase[],
+  user: string,
+  relayer: string
+): [SwapAction, number] {
+  const actionStep = getActionStep(
+    mainTokenInIndex,
+    mainTokenOutIndex,
+    swap.assetInIndex,
+    swap.assetOutIndex
+  );
+  // Will get actual amount if input or chain amount if part of chain
+  const amountIn = getActionAmount(
+    swap.amount,
+    ActionType.Swap,
+    actionStep,
+    opRefKey
+  );
+  // Updates swap data to use chainedRef if required
+  swap.amount = amountIn;
+  // This will be 0 if not a mainTokenOut action otherwise amount using slippage
+  const minOut = getActionMinOut(swap.returnAmount ?? '0', slippage);
+  // This will set opRef for next chained action if required
+  const [opRef, newOpRefKey] = getActionOutputRef(
+    actionStep,
+    swap.assetOutIndex,
+    opRefKey
+  );
+  const hasTokenIn =
+    actionStep === ActionStep.Direct || actionStep === ActionStep.TokenIn
+      ? true
+      : false;
+  const hasTokenOut =
+    actionStep === ActionStep.Direct || actionStep === ActionStep.TokenOut
+      ? true
+      : false;
+  const isBptIn = isBpt(pools, assets[swap.assetInIndex]);
+  // joins - can't join a pool and send BPT to internal balances
+  // Because of ^ we can assume that any BPT is coming from external (either from user or join)
+  let fromInternal = true;
+  if (hasTokenIn || isBptIn) fromInternal = false;
```

Comment:
you can replace the 2 lines above by:
const fromInternal = !hasTokenIn && !isBptIn // or  !(hasTokenIn || isBptIn)  but the first is more performative